pkg/pillar/cmd/downloader/syncop.go

```diff
@@ -0,0 +1,272 @@
+package downloader
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"os"
+	"time"
+
+	zconfig "github.com/lf-edge/eve/api/go/config"
+	"github.com/lf-edge/eve/pkg/pillar/types"
+	"github.com/lf-edge/eve/pkg/pillar/zedUpload"
+	"github.com/lf-edge/eve/pkg/pillar/zedcloud"
+	log "github.com/sirupsen/logrus"
+)
+
+// Drona APIs for object Download
+func handleSyncOp(ctx *downloaderContext, key string,
+	config types.DownloaderConfig, status *types.DownloaderStatus,
+	dst *types.DatastoreConfig) {
+	var (
+		err                                        error
+		errStr, locFilename, remoteName, serverURL string
+		syncOp                                     zedUpload.SyncOpType = zedUpload.SyncOpDownload
+		trType                                     zedUpload.SyncTransportType
+		auth                                       *zedUpload.AuthInput
+	)
+
+	if status.ObjType == "" {
+		log.Fatalf("handleSyncOp: No ObjType for %s\n",
+			status.Safename)
+	}
+
+	log.Debugf("handleSyncOp: config: %+v", config)
+	log.Debugf("handleSyncOp: IsContainer: %v", config.IsContainer)
+
+	// get the datastore context
+	dsCtx := constructDatastoreContext(config, status, dst)
+
+	// by default the metricsURL _is_ the DownloadURL, but can override in switch
+	metricsUrl := dsCtx.DownloadURL
+
+	if config.IsContainer {
+		pullPolicy := "new"
+		rktFetchContainerImage(ctx, key, config, status, *dsCtx, pullPolicy)
+		return
+	}
+	locDirname := types.DownloadDirname + "/" + status.ObjType
+	locFilename = locDirname + "/pending"
+
+	// update status to DOWNLOAD STARTED
+	status.State = types.DOWNLOAD_STARTED
+	publishDownloaderStatus(ctx, status)
+
+	if config.ImageSha256 != "" {
+		locFilename = locFilename + "/" + config.ImageSha256
+	}
```

Comment:
What happens in the else case? Can we get a config with Empty SHA? What do we do in that case? Should we return an Error? Otherwise, we will just use the name locDirname/pending .. And if there are multiple such simultaneous downloads, we will run into errors..

Better return error check ( May be early in the process ) or just log.Fatal..

Reply:
Yes, that is what will happen, which is not good. However, this is another one of those, "this is how it worked before". Yes, it should be fixed, but let's iterate on it first. Separation here, fix those errors in several follow-ups. As long as we are not making it any worse, let's move ahead.

> Better return error check ( May be early in the process ) or just log.Fatal

Part of the issue is that `handleSyncOp()` doesn't return _anything_, instead it either does `log.Fatalf()`, or it calls `handleSyncOpResponse()` (or both). Using `log.Fatalf()` is not a good move here (best handled only in main functions, or places that are meant to work with them), but, again, one step at a time.